### PR TITLE
Dev pm 5371; Radau5 refactorization

### PR DESCRIPTION
**Pull request by PeterMeisrimelModelon**

### CODE REVIEW

**Review by chria — COMMENTED** [review]
Looks good! Just a few minor comments

File: src/solvers/radau5.py [hunk 1]
```diff
@@ -46,22 +47,22 @@ class Radau5Error(AssimuloException):
...
-            -6    : 'Failure in sparse Jacobian evaluation, specified number of nonzero elements too small.',
-            -7    : 'Sparse Jacobian given in wrong format, expects CSC format.',
-            -8    : 'Unexpected internal function call failure of SUPERLU.',
-            -9    : 'Memory allocation failure in SUPERLU.',
-            -10   : 'Unrecoverable exception encountered during callback to problem (right-hand side/jacobian).',
+            -10   : 'Unrecoverable exception encountered during callback to problem (right-hand side/jacobian).'
             }
     
-    def __init__(self, value, t = 0.0):
+    def __init__(self, value: int = None, t: float = 0.0, err_msg: str = None):
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/solvers/radau5.py]: Good point, removed these.

File: src/solvers/radau5.py [hunk 2]
```diff
@@ -93,24 +94,22 @@ def __init__(self, problem):
         """
         Explicit_ODE.__init__(self, problem) #Calls the base class
         
-        #Default values
+        #Default values; None = Radau5 decides
         self.options["inith"]    = 0.01
         self.options["newt"]     = 7 #Maximum number of newton iterations
         self.options["thet"]     = 1.e-3 #Boundary for re-calculation of jac
-        self.options["fnewt"]    = 0.0 #Stopping critera for Newtons Method
+        self.options["fnewt"]    = None #Stopping critera for Newtons Method
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/solvers/radau5.py:101]: fnewt depends on the relative tolerance and uround. As such, setting a fixed value needs to be done with great care and we instead use the default value.
Note: All other inputs (except maxh; default: maxh = tend - t0) are independent on other inputs/parameters.

Previously, all these options were pasted into the work array and a value of 0.0 meant the Radau5 solver would choose. I think "None" is a more clear indicator that a given option is not set.

File: src/solvers/radau5.py [hunk 3]
```diff
@@ -93,24 +94,22 @@ def __init__(self, problem):
...
         self.options["quot1"]    = 1.0 #Parameters for changing step-size (lower bound)
         self.options["quot2"]    = 1.2 #Parameters for changing step-size (upper bound)
         self.options["fac1"]     = 0.2 #Parameters for step-size selection (lower bound)
         self.options["fac2"]     = 8.0 #Parameters for step-size selection (upper bound)
-        self.options["maxh"]     = N.inf #Maximum step-size.
+        self.options["maxh"]     = None #Maximum step-size.
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/solvers/radau5.py:106]: Addressed in comment above

File: src/solvers/radau5.py [hunk 4]
```diff
@@ -148,40 +147,13 @@ def _set_linear_solver(self, linear_solver):
     linear_solver = property(_get_linear_solver, _set_linear_solver)
 
     def _get_implementation(self):
-        return self.options["implementation"]
-    
-    def _set_implementation(self, implementation):
+        return 'c'
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/solvers/radau5.py:151]: Good point. Added

self.log_message("Deprecation Warning: Radau5ODE only supports the 'c' implementation and this attribute will be removed in the future\n", LOUD)

File: src/solvers/radau5.py [hunk 5]
```diff
@@ -148,40 +147,13 @@ def _set_linear_solver(self, linear_solver):
...
-            try:
-                from assimulo.lib import radau5_c_py as radau5_c
-                self.radau5 = radau5_c
-                self.solver_module_imported = True
-            except Exception:
-                raise Radau_Exception("Failed to import the C based Radau5 solver implementation. Note that this solver requires an installation with SuperLU. Try using 'implementation' = 'f' for the Fortran based solver instead.") from None
-        else:
-            raise Radau_Exception("'implementation' parameter needs to be either 'f' or 'c'. Set value: {}".format(implementation)) from None
-        self.options["implementation"] = implementation_lower
+        self.log_message("Warning: Radau5ODE only supports the 'c' implementation.\n", LOUD)
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/solvers/radau5.py]: I took your suggestion.

File: thirdparty/radau5/radau5ode.pyx [hunk 6]
```diff
@@ -175,99 +232,49 @@ cpdef radau5(fcn_PY, double x, np.ndarray y,
...
                         - Final time for which a solution has been computed, x == xend, if succesful
             y
                         - Final solution
-            h__
-                        - Prediced size of last accepted step
-            iwork
-                        - Statistics about number of function calls etc, see radau_decsol.c for details
             idid
-                        - Return flag, see radau_decsol.c for details (1 == Successful computation)
+                        - Return flag, see radau5_c.c 0 = sucess, 1 = sucess, interrput by solout
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply thirdparty/radau5/radau5ode.pyx]: Fixed.

File: thirdparty/radau5/radau5_io.c [hunk 7]
```diff
@@ -0,0 +1,535 @@
...
+		#ifdef __RADAU5_WITH_SUPERLU
+			if (nnz < 0){
+				sprintf(rmem->err_log, "Input nnz must be nonnegative, received nnz = %i", nnz);
+				return RADAU_ERROR_INCONSISTENT_INPUT;
+			}
+			if (nprocs < 1){
+				sprintf(rmem->err_log, "Input nprocs must be positive, received nprocs = %i", nprocs);
+				return RADAU_ERROR_INCONSISTENT_INPUT;
+			}
+			rmem->lin_sol->nnz = nnz > n*n ? nnz : nnz; /* automatically cap at n*n */
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply thirdparty/radau5/radau5_io.c]: Good catch, fixed.

File: thirdparty/radau5/radau5_c.c [hunk 8]
```diff
@@ -0,0 +1,1293 @@
+/* based on f2c (version 20100827) translation of radau_decsol.f. */
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply thirdparty/radau5/radau5_c.c]: Fixed.

File: src/solvers/radau5.py [hunk 9]
```diff
@@ -46,22 +47,22 @@ class Radau5Error(AssimuloException):
...
+        self.err_msg = err_msg
         
-    def __str__(self): 
-        try:
-            return repr(self.msg[self.value]+' At time %f.'%self.t)    
-        except KeyError:
-            return repr('Radau failed with flag %s. At time %f.'%(self.value, self.t))
+    def __str__(self):
+        if self.err_msg:
+            return repr('Radau failed with flag %s. At time %f. Message: %s'%(self.value, self.t, self.err_msg))
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/solvers/radau5.py]: Done, had to update tests accordingly

File: src/solvers/radau5.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chria) [reply src/solvers/radau5.py:101]: Ok I see. Thanks for the clarification.

**Pull request merged**